### PR TITLE
Add MessagePack serialization & function decorator

### PR DESCRIPTION
**Pull request by yanndupis**
This PR add:
- Serialization/deserialization with with [MessagePack](https://github.com/msgpack/msgpack-python) for   
   - python native types: `string`, `int`, `float`, `complex`,  `list`, `dict`, `set`, `frozenset`, `tuple`, `bool`, `bytes` and `bytearray`.
   - numpy arrray for easy experimentation for ML/DS use case.
- `io_serialize` (an idea for a better name is more than welcome ? :)) decorator to deserialize input of a cape function and serialize output.
- Add `msgpack_serialize` argument to `cape.run` to `cape.invoke`  to let the user decide if he/she wants to have PyCape serialize the input with MessagePack automatically. If set to False `cape.run` and `cape.invoke` expect to have the input already serialized as bytes. Should we set `msgpack_serialize` as True by default or False for now? if set to True as default it means we expect the user to the decorator for the cape function as default to serialize/deserialize the same way as `cape.run` and `cape.invoke`.
- Currently support single input and output but could be extended if we decide to support multiple inputs/outputs to `cape.run`,  `cape.invoke` and Cape functions.

Additional notes:
- To use  the `io_serialize` decorator, the user will have to add PyCape as dependency when deploying the function to be able to run the function within the enclave as dependency. What do we think about having pycape already install in the enclave so the user doesn't have to add it as dependency manually even for simple functions.
- Using MessagePack serialization and the decorator, we should be able to deserialize input sent from the Javascript SDK by handling properly the javascript type in the `decode` function in the `serialize` file.

[JIRA-666](https://capeprivacy.atlassian.net/browse/CAPE-666?atlOrigin=eyJpIjoiNWU0NTViMzQyZjQ1NDM3OWFiNzFlMTUxNDliNDkxNTQiLCJwIjoiaiJ9)

### CODE REVIEW

File: examples/invoke.py [hunk 1]
```diff
@@ -11,13 +11,13 @@
     )
     cape.connect(function_id)
 
-    result = cape.invoke("Hello Cape")
+    result = cape.invoke("Hello Cape".encode(), msgpack_serialize=False)
```
Comment (bendecoste) [inline examples/invoke.py]: Do we need this since it defaults to false?

File: pycape/cape.py [hunk 2]
```diff
@@ -61,24 +71,36 @@ async def _connect(self, function_id):
...
-        input_bytes = _convert_input_to_bytes(input)
-        ciphertext = encrypt(input_bytes, self._public_key)
+    async def _invoke(self, input, msgpack_serialize=False):
+        if not isinstance(input, bytes):
+            if msgpack_serialize:
+                input = serialize(input)
+            else:
+                raise ValueError(
+                    f"The input type is: {type(input)}. Provide input as bytes or "
+                    "set msgpac_serialize in cape.run or cape.invoke as True to "
```
Comment (bendecoste) [inline pycape/cape.py]: should say `msgpack`

**Review by bendecoste — COMMENTED** [review]
Are you able to make an example that does use the message pack serialization? I think that might make it easier to see how everything fits together

**Comment by yanndupis** [comment]
> Are you able to make an example that does use the message pack serialization? I think that might make it easier to see how everything fits together

Good point, I will add this tomorrow. In the meantime you can take a look at the example I just added to the [functions repo](https://github.com/capeprivacy/functions/pull/3/files). Here is the [app.py file](https://github.com/capeprivacy/functions/blob/49125ec86f368431d206128c0c4b0abaefba0781/np-stats/app.py) and  example of [run file](https://github.com/capeprivacy/functions/blob/49125ec86f368431d206128c0c4b0abaefba0781/np-stats/run.py). Thanks

**Comment by yanndupis** [comment]
Hey @bendecoste, I have added an example. It should be ready for review. thank you!

**Review by justin1121 — DISMISSED** [review]

**Review by justin1121 — APPROVED** [review]

**Pull request merged**